File: src/class039/C3_NumberOfAtoms.py
# 含有嵌套的分子式求原子数量
# 测试链接 : https://leetcode.cn/problems/number-of-atoms/
class Code03_NumberOfAtoms:
    def __init__(self):
        self.where = 0

    # 主函数
    def count_of_atoms(self, formula: str) -> str:
        self.where = 0
        atom_map = self._parse_formula(formula, 0)
        result = []
        for atom, count in sorted(atom_map.items()):
            result.append(atom)
            if count > 1:
                result.append(str(count))
        return ''.join(result)

    # formula[i....]开始计算，遇到字符串终止 或者 遇到 ')' 停止
    # 返回 : 自己负责的这一段字符串的结果，有序表！
    # 返回之间，更新全局变量where，为了上游函数知道从哪继续！
    def _parse_formula(self, formula: str, i: int):
        atom_map = {}
        name = ''
        pre = {}
        cnt = 0
        while i < len(formula) and formula[i] != ')':
            if formula[i].isupper() or formula[i] == '(':
                self._fill(atom_map, name, pre, cnt)
                name = ''
                pre = {}
                cnt = 0
                if formula[i].isupper():
                    name += formula[i]
                    i += 1
                else:  # 遇到 '('
                    pre, i = self._parse_formula(formula, i + 1)
            elif formula[i].islower():
                name += formula[i]
                i += 1
            else:
                cnt = cnt * 10 + int(formula[i])
                i += 1
        self._fill(atom_map, name, pre, cnt)
        self.where = i
        return atom_map, self.where

    # 将收集到的数据填充到总表中
    def _fill(self, atom_map, name, pre, cnt):
        cnt = 1 if cnt == 0 else cnt
        if name:
            atom_map[name] = atom_map.get(name, 0) + cnt
        else:
            for atom, count in pre.items():
                atom_map[atom] = atom_map.get(atom, 0) + count * cnt

File: src/class039/C3_NumberOfAtoms.cpp
#include <string>
#include <map>
using namespace std;

// 含有嵌套的分子式求原子数量
// 测试链接 : https://leetcode.cn/problems/number-of-atoms/
class Code03_NumberOfAtoms {
public:
    static int where;

    // 主函数
    static string countOfAtoms(string str) {
        where = 0;
        map<string, int> map = f(str, 0);
        string ans;
        for (auto &entry : map) {
            ans += entry.first;
            int cnt = entry.second;
            if (cnt > 1) {
                ans += to_string(cnt);
            }
        }
        return ans;
    }

    // s[i....]开始计算，遇到字符串终止 或者 遇到 ')' 停止
    // 返回 : 自己负责的这一段字符串的结果，有序表！
    // 返回之间，更新全局变量where，为了上游函数知道从哪继续！
    static map<string, int> f(const string &s, int i) {
        map<string, int> ans;
        string name;
        map<string, int> pre;
        int cnt = 0;
        while (i < s.length() && s[i] != ')') {
            if ((s[i] >= 'A' && s[i] <= 'Z') || s[i] == '(') {
                fill(ans, name, pre, cnt);
                name.clear();
                pre.clear();
                cnt = 0;
                if (s[i] >= 'A' && s[i] <= 'Z') {
                    name += s[i++];
                } else { // 遇到 '('
                    pre = f(s, i + 1);
                    i = where + 1;
                }
            } else if (s[i] >= 'a' && s[i] <= 'z') {
                name += s[i++];
            } else {
                cnt = cnt * 10 + s[i++] - '0';
            }
        }
        fill(ans, name, pre, cnt);
        where = i;
        return ans;
    }

    // 将收集到的数据填充到总表中
    static void fill(map<string, int> &ans, string &name, map<string, int> &pre, int cnt) {
        if (!name.empty() || !pre.empty()) {
            cnt = cnt == 0 ? 1 : cnt;
            if (!name.empty()) {
                ans[name] += cnt;
            } else {
                for (auto &entry : pre) {
                    ans[entry.first] += entry.second * cnt;
                }
            }
        }
    }
};

int Code03_NumberOfAtoms::where = 0;
